File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.js
/**
 * @param {number[][]} classes
 * @param {number} extraStudents
 * @return {number}
 */
var maxAverageRatio = function(classes, extraStudents) {
  const q = new MaxPriorityQueue({
    priority: i => (classes[i][0] + 1) / (classes[i][1] + 1) - 
      classes[i][0] / classes[i][1],
  });
  classes.forEach((_, i) => {
    q.enqueue(i);
  });
  
  let es = extraStudents;
  while (es) {
    const i = q.dequeue().element;
    
    classes[i][0] += 1;
    classes[i][1] += 1;
    q.enqueue(i);
    
    es -= 1;
  }
  
  let sum = 0;
  classes.forEach(([p, t]) => {
    sum += p / t;
  });
  
  return sum / classes.length;
};

File: 1792-maximum-average-pass-ratio/1792-maximum-average-pass-ratio.py
class Solution(object):
    def maxAverageRatio(self, classes, extraStudents):
        def profit(a, b):
            return (a + 1) / (b + 1) - a / b

        maxHeap = []
        for a, b in classes:
            a, b = a * 1.0, b * 1.0  # Convert int to double
            maxHeap.append((-profit(a, b), a, b))
        heapq.heapify(maxHeap)  # Heapify maxHeap cost O(N)

        for _ in range(extraStudents):
            d, a, b = heapq.heappop(maxHeap)
            heapq.heappush(maxHeap, (-profit(a + 1, b + 1), a + 1, b + 1))

        return sum(a / b for d, a, b in maxHeap) / len(classes)